### PR TITLE
add filter option to deps.unlock

### PR DESCRIPTION
**Pull request by mobileoverlord**

### CODE REVIEW

File: lib/mix/lib/mix/tasks/deps.unlock.ex [hunk 1]
```diff
@@ -29,6 +29,26 @@ defmodule Mix.Tasks.Deps.Unlock do
...
+          |> Enum.filter(&String.contains?(&1, filter))
+
+        if unlock == [] do
+          Mix.shell.error "warning: no dependency were matched"
+        else
+          lock = Enum.reject(apps, & &1 in unlock)
+          Mix.Dep.Lock.write(lock)
+          Mix.shell.info """
+          Unlocked deps
+          #{Enum.join(unlock, "\n")}
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/deps.unlock.ex]: * #{Enum.join(unlock, "\n * ")}

File: lib/mix/lib/mix/tasks/deps.unlock.ex [hunk 2]
```diff
@@ -29,6 +29,26 @@ defmodule Mix.Tasks.Deps.Unlock do
...
+
+        unlock =
+          apps
+          |> Enum.map(&Atom.to_string/1)
+          |> Enum.filter(&String.contains?(&1, filter))
+
+        if unlock == [] do
+          Mix.shell.error "warning: no dependency were matched"
+        else
+          lock = Enum.reject(apps, & &1 in unlock)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/deps.unlock.ex]: `apps -- unlock`

File: lib/mix/lib/mix/tasks/deps.unlock.ex [hunk 3]
```diff
@@ -29,6 +29,26 @@ defmodule Mix.Tasks.Deps.Unlock do
       opts[:unused] ->
         apps = Mix.Dep.loaded([]) |> Enum.map(& &1.app)
         Mix.Dep.Lock.read() |> Map.take(apps) |> Mix.Dep.Lock.write()
+      opts[:filter] ->
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/deps.unlock.ex]: `filter = opts[:filter] ->`

File: lib/mix/lib/mix/tasks/deps.unlock.ex [hunk 4]
```diff
@@ -29,6 +29,26 @@ defmodule Mix.Tasks.Deps.Unlock do
...
+          |> Enum.map(&Atom.to_string/1)
+          |> Enum.filter(&String.contains?(&1, filter))
+
+        if unlock == [] do
+          Mix.shell.error "warning: no dependency were matched"
+        else
+          lock = Enum.reject(apps, & &1 in unlock)
+          Mix.Dep.Lock.write(lock)
+          Mix.shell.info """
+          Unlocked deps
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/deps.unlock.ex]: `Unlocked deps:`

**Comment by josevalim** [comment]
I have added some comments and then we just need a test. :)

**Comment by mobileoverlord** [comment]
I added a test and went keps reject since we needed to filter the original lock map

**Comment by josevalim** [comment]
Do you need this backported?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by mobileoverlord** [comment]
Please backport. Thanks!

**Pull request merged**